baseplate/server/queue_consumer.py

```diff
@@ -164,23 +139,62 @@ def run(self) -> None:
-                # new message.  If we did not do this, we would not be able to
-                # wait for all of our workers to finish before stopping the
-                # server.
-                message = self.work_queue.get(timeout=self._queue_timeout)
-            except queue.Empty:
-                pass
-            else:
-                # Ensure that if self.message_handler.handle throws a `queue.Empty`
-                # error, that bubbles up and is not treated as though `self.work_queue`
-                # is empty
-                self.message_handler.handle(message)
+            self._get_and_handle_message()
         logger.debug("Consumer <%s> stopping.", self.id)
 
+    def _get_and_handle_message(self) -> None:
+        """Get and process a message from the queue if it is not empty."""
+        try:
+            # We set a timeout so we can periodically check if we should
+            # stop, this way we will actually return if we have recieved a
+            # `stop()` call rather than hanging in a `get` waiting for a
+            # new message.  If we did not do this, we would not be able to
+            # wait for all of our workers to finish before stopping the
+            # server.
+            message = self.work_queue.get(timeout=self._queue_timeout)
+        except queue.Empty:
+            pass
+        else:
+            # Ensure that if self.message_handler.handle throws a `queue.Empty`
+            # error, that bubbles up and is not treated as though `self.work_queue`
+            # is empty
+            self._handle_message(message)
+
+    def _handle_message(self, message: Any) -> None:
+        """Handle a single message."""
+        return self.message_handler.handle(message)
+
+
+class QueueConsumerFactory(abc.ABC):
+    """Factory for building all of the objects needed to run a QueueConsumerServer.
+
+    This is the type of object that you build in your `service.make_queue_consumer`
+    function to be passed to `make_server`.
+    """
+
+    @abc.abstractmethod
+    def build_pump_worker(self, work_queue: queue.Queue) -> PumpWorker:
+        """Build an object implementing the PumpWorker interface.
+
+        `work_queue` is the Queue that will be shared between the PumpWorker and
+        and the QueueConsumer to pass messages to the MessageHandler, be sure
+        that your PumpWorker uses `work_queue`.
+        """
+
+    @abc.abstractmethod
+    def build_message_handler(self) -> MessageHandler:
+        """Build an object implementing the MessageHandler interface."""
+
+    @abc.abstractmethod
+    def build_health_checker(self, listener: socket.socket) -> StreamServer:
+        """Build an HTTP server to service health checks."""
+
+    def build_queue_consumer(
+        self, work_queue: queue.Queue, message_handler: MessageHandler
+    ) -> QueueConsumer:
+        """Build a queue consumer."""
+        return QueueConsumer(work_queue=work_queue, message_handler=message_handler)
```

Comment:
To go along with my comment about not needing to extend the base `QueueConsumer` I don't think these changes are really necessary either?

Reply:
removed